parquet/src/main/java/org/apache/iceberg/parquet/ParquetSchemaUtil.java

```diff
@@ -87,29 +82,21 @@ public static boolean hasIds(MessageType fileSchema) {
       // Try to convert the type to Iceberg. If an ID assignment is needed, return false.
       ParquetTypeVisitor.visit(fileSchema, new MessageTypeToType(fileSchema) {
         @Override
-        protected int nextId() {
-          throw new IllegalStateException("Needed to assign ID");
+        protected int getId(org.apache.parquet.schema.Type type) {
+          org.apache.parquet.schema.Type.ID id = type.getId();
+          if (id != null) {
+            throw new IllegalStateException("at least one ID exists");
```

Comment:
It is fine if not everything has an ID. Columns without IDs will be ignored and should not cause a read to fail.

Reply:
Did I misunderstand the comment?   I thought you meant file schema either has all columns with IDs or do not. Do you mean that a file schema can have some of the columns with IDs and some without ID?

> This also needs to update ParquetSchemaUtil.hasIds so that any ID causes it to return true, instead of any missing ID. This should be okay since files either have IDs or do not, so it is not a behavior change.

Reply:
Yes, we expect schemas to either have all IDs or none, but we don't need to fail if that assumption is violated. If we detect any ID in the schema, then we should trust the file's IDs. If there are no IDs, then we should infer them using a name mapping or position-based fallback, whatever was configured.